### DISCUSSION TITLE
Shader issue with 2.0 and M1 Mac

### DISCUSSION THREAD

**Discussion by chadwyk** [axmolengine/axmol#1329] [Q&A] [answered]
Hi, I downloaded and installed both the 2.0 release and the latest Dev release and am having issues that I didn't have in 1.0.

I have tried building and running both a new project, and cpp_tests. The apps build using an Xcode project for MacOS fine without error. As soon as the app launches it crashes with the following:

Console: IPV4 server is listening on 0.0.0.0:5678
Loading shader: 5 positionTextureColor_vs, positionTextureColor_fs ...
axmol: fullPathForFilename: No file found at positionTextureColor_vs. Possible missing file.
axmol: fullPathForFilename: No file found at positionTextureColor_fs. Possible missing file.
libc++abi: terminating due to uncaught exception of type std::out_of_range: ibstream_view::consume out of range!
Program ended with exit code: 9

I do see that positionTextureColor_vs and positionTextureColor_fs do exist in build/runtime/axslc.  It looks like this folder is new as my 1.0 project does not create this folder when it builds.  Does this folder need to be added to the project someplace in the build settings? 

**Resolved**: Since v2 requires a special version of cmake that it downloads on setup, to build projects use "axmol build -p osx -a arm64 -c" instead of "cmake -S . -B build -GXcode -DCMAKE_OSX_ARCHITECTURES=arm64"

**Comment by halx99** [comment]
please run setup.ps1, axmol2 require a unreleased cmake version

**Reply by chadwyk** [reply]
Yup, I did when I installed it. I ran "pwsh setup.ps1" I noticed that when running, it did download cmake, no errors. I quit the terminal and reopened as instructed. Then followed the normal routine in the documentation to create a new project.

I just deleted everything and tried yet again, same issue.

**Comment by halx99** [comment]
can you post the setup and cmake generate log?

axmol build -p osx -a arm64 -c

I have no m1/m2 device, so i can't verify it.

**Reply by halx99** [reply]
and open xcodeproj by xcode, check whether the folder axslc in Embed Resources group

**Reply by chadwyk** [reply]
Where do I find this in the project? Is it in the build settings? Or do you mean just in the left side where it lists all the source files?

**Comment by chadwyk** [comment]
I assume you just want what is listed in the terminal when running that command? See below:

chadwyk@static-211-244 axmol-dev % axmol build -p osx -a arm64 -c
fatal: not a git repository (or any of the parent directories): .git
-p osx -a arm64 -c --d: /Users/chadwyk/xcode_projects/axmol-dev
build1k: PowerShell 7.3.6
build1k: 
Name                           Value
----                           -----
sdk                            
a                              arm64
p                              osx
d                              /Users/chadwyk/xcode_projects/axmol-dev
xc                             {}
dll                            False
xb                             {--target, HelloCpp, --config, Debug}
xt                             cmake
cc                             
prefix                         /Users/chadwyk/xcode_projects/axmol-dev/tools/ex…


build1k: proj_dir=/Users/chadwyk/xcode_projects/axmol-dev, prefix=/Users/chadwyk/xcode_projects/axmol-dev/tools/external
build1k: Using glslcc: /Users/chadwyk/xcode_projects/axmol-dev/tools/external/glslcc/glslcc, version: 1.9.3
build1k: Building target osx on macos with toolchain xcode ...
build1k: CONFIG_ALL_OPTIONS=-GXcode -DCMAKE_OSX_ARCHITECTURES=arm64, Count=2
-- The C compiler identification is AppleClang 14.0.3.14030022
-- The CXX compiler identification is AppleClang 14.0.3.14030022
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Check for working C compiler: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/clang - skipped
-- Detecting C compile features
-- Detecting C compile features - done
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Check for working CXX compiler: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/clang++ - skipped
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- GLSLCC_OUT_DIR=/Users/chadwyk/xcode_projects/axmol-dev/build_arm64/runtime/axslc
-- GLSLCC_FIND_PROG_ROOT=/Users/chadwyk/xcode_projects/axmol-dev/tools/external/glslcc
-- GLSLCC_FRAG_SOURCE_FILE_EXTENSIONS=.frag;.fsh
-- GLSLCC_VERT_SOURCE_FILE_EXTENSIONS=.vert;.vsh
-- CMAKE_C_STANDARD=11
-- CMAKE_CXX_STANDARD=20
-- The ASM compiler identification is AppleClang
-- Found assembler: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/clang
-- The ASM_NASM compiler identification is unknown
-- Didn't find assembler
CMake Warning at cmake/Modules/AXConfigDefine.cmake:215 (message):
  The nasm compiler doesn't present on your system PATH, please download
  from: https://www.nasm.us/pub/nasm/releasebuilds/2.16.01/
Call Stack (most recent call first):
  cmake/Modules/AXBuildSet.cmake:47 (include)
  CMakeLists.txt:46 (include)


-- CMAKE_HOST_SYSTEM_NAME:Darwin
-- CMAKE_SYSTEM_NAME:Darwin
-- CMAKE_GENERATOR_PLATFORM:
-- CMAKE_SYSTEM_PROCESSOR:arm64
-- CMAKE_CXX_COMPILER_ID:AppleClang
-- FUZZ_MSVC=, FULL_CLANG=
-- PROJECT_NAME:axmol
-- PROJECT_SOURCE_DIR:/Users/chadwyk/xcode_projects/axmol-dev
-- _AX_ROOT:/Users/chadwyk/xcode_projects/axmol-dev
-- CMAKE_MODULE_PATH:/Users/chadwyk/xcode_projects/axmol-dev/cmake/Modules/
-- PROJECT_BINARY_DIR:/Users/chadwyk/xcode_projects/axmol-dev/build_arm64
-- ENGINE_BINARY_PATH:/Users/chadwyk/xcode_projects/axmol-dev/build_arm64/engine
-- ARCH_ALIAS:
CMake Warning (dev) at core/CMakeLists.txt:87 (message):
  AX_USE_GL=OFF, AX_USE_METAL=ON, AX_GLES_PROFILE=0
This warning is for project developers.  Use -Wno-dev to suppress it.

-- AX_ENABLE_VLC_MEDIA=OFF
-- Found Git: /usr/bin/git (found version "2.39.2 (Apple Git-143)") 
-- Performing Test AX_HAVE_SSE2_SWITCH
-- Performing Test AX_HAVE_SSE2_SWITCH - Failed
-- Performing Test AX_HAVE_AVX2_INTRINSICS
-- Performing Test AX_HAVE_AVX2_INTRINSICS - Failed
-- Performing Test AX_HAVE_SSE42_INTRINSICS
-- Performing Test AX_HAVE_SSE42_INTRINSICS - Failed
-- Performing Test AX_HAVE_SSE41_INTRINSICS
-- Performing Test AX_HAVE_SSE41_INTRINSICS - Failed
-- Performing Test AX_HAVE_SSE2_INTRINSICS
-- Performing Test AX_HAVE_SSE2_INTRINSICS - Failed
-- Looking for arm_neon.h
-- Looking for arm_neon.h - found
-- Performing Test AX_HAVE_NEON_INTRINSICS
CMake Warning (dev) at thirdparty/CMakeLists.txt:129 (message):
  
  AX_ISA_SIMD=neon,AX_HAVE_AVX2_INTRINSICS=,AX_HAVE_SSE42_INTRINSICS=,AX_HAVE_SSE41_INTRINSICS=,AX_HAVE_SSE2_INTRINSICS=,AX_HAVE_NEON_INTRINSICS=1
This warning is for project developers.  Use -Wno-dev to suppress it.

-- Performing Test AX_HAVE_NEON_INTRINSICS - Success
-- Using plainlua https://lua.org
-- Version: 10.1.0
-- Build type: 
-- Performing Test HAS_NULLPTR_WARNING
-- Performing Test HAS_NULLPTR_WARNING - Success
Configuring Chipmunk2D version 7.0.3
CMake Deprecation Warning at thirdparty/freetype/CMakeLists.txt:113 (cmake_minimum_required):
  Compatibility with CMake < 3.5 will be removed from a future version of
  CMake.

  Update the VERSION argument <min> value or use a ...<max> suffix to tell
  CMake that the project does not need compatibility with older versions.


-- Found PkgConfig: /opt/homebrew/bin/pkg-config (found version "0.29.2") 
-- Found ZLIB: /Users/chadwyk/xcode_projects/axmol-dev/thirdparty/zlib/prebuilt/mac/libz.a (found version "1.3.#define ZLIB_VERSION "1.3"")  
-- Found PNG: png (found version "1.6.40") 
-- Looking for unistd.h
-- Looking for unistd.h - found
-- Looking for fcntl.h
-- Looking for fcntl.h - found
-- Performing Test ASTC_HAVE_NEON_INTRINSICS
CMake Warning (dev) at thirdparty/astcenc/CMakeLists.txt:89 (message):
  ASTC_ISA_SIMD=neon
This warning is for project developers.  Use -Wno-dev to suppress it.

-- Performing Test ASTC_HAVE_NEON_INTRINSICS - Success
-- Performing Test CMAKE_HAVE_LIBC_PTHREAD
-- Performing Test CMAKE_HAVE_LIBC_PTHREAD - Success
-- Found Threads: TRUE  
-- Including Cocoa support
-- AX_USE_ALSOFT=OFF
CMake axmol target_precompile_headers
-- Enabled axmol extensions:cocostudio;Inspector;ImGui;fairygui;DragonBones;spine;physics-nodes;particle3d;assets-manager;GUI;
-- CMAKE_BINARY_DIR=/Users/chadwyk/xcode_projects/axmol-dev/build_arm64
-- No shader found in /Users/chadwyk/xcode_projects/axmol-dev/templates/cpp-template-default/Source/shaders
-- 50 shader sources found in /Users/chadwyk/xcode_projects/axmol-dev/tests/cpp-tests/Source/shaders
CMake cpp_tests target_precompile_headers
CMake fairygui_tests target_precompile_headers
-- No shader found in /Users/chadwyk/xcode_projects/axmol-dev/tests/fairygui-tests/Source/shaders
-- No shader found in /Users/chadwyk/xcode_projects/axmol-dev/templates/lua-template-default/Source/shaders
-- No shader found in /Users/chadwyk/xcode_projects/axmol-dev/tests/lua-tests/Source/shaders
-- Configuring done (17.9s)
-- Generating done (0.0s)
-- Build files have been written to: /Users/chadwyk/xcode_projects/axmol-dev/build_arm64
build1k: Build done

**Reply by halx99** [reply]
/Users/chadwyk/xcode_projects/axmol-dev/build_arm64

**Reply by chadwyk** [reply]
Yup, that's the project file in /Users/chadwyk/xcode_projects/axmol-dev/build (not build_arm64) I've looked all over for axslc in the settings or folders once it opens in Xcode.  I googled it a bit, I tried searching for the flag XCODE_EMBED_RESOURCES but didn't find that at all either.

**Reply by halx99** [reply]
but your cmake print the build folder is build_arm64
-- Generating done (0.0s)
-- Build files have been written to: /Users/chadwyk/xcode_projects/axmol-dev/build_arm64

**Reply by chadwyk** [reply, marked as answer]
Ah sorry, yes... I think this is the root of the issue.... After running "axmol build -p osx -a arm64 -c" it does put it in "build_arm64" and I am able to build cpp_tests via the Xcode app!

I had been also using "cmake -S . -B build -GXcode -DCMAKE_OSX_ARCHITECTURES=arm64" (found in the documentation for  creating MacOS project) which places it just in the build folder.  That does not work.

I just tested it with a new Mac project with the command you gave me, and that works too.

So it looks like the documentation needs to be updated: https://github.com/axmolengine/axmol/blob/dev/docs/DevSetup.md

**Reply by chadwyk** [reply]
My assumption maybe the old command that I was running was using the old version of cmake found in my Applications folder rather than the version that Axmol's setup script downloaded?

**Reply by halx99** [reply]
Yes, the axmol build commands will auto setup correct depends tools.

**Reply by halx99** [reply]
done: https://github.com/axmolengine/axmol/blob/dev/docs/DevSetup.md#ios-tvos-and-macos

**Comment by halx99** [comment]
i haven't see the cmake version, it's important and the github action print: `build1k: Using cmake: /Users/runner/work/axmol/axmol/tools/external/cmake/bin/cmake, version: 3.27.20230808-g94eae40`

**Reply by chadwyk** [reply]
chadwyk@static-211-244 axmol-dev %  pwsh setup.ps1   
build1k: PowerShell 7.3.6
build1k: proj_dir=/Users/chadwyk/xcode_projects/axmol-dev, prefix=/Users/chadwyk/xcode_projects/axmol-dev/tools/external
build1k: Downloading https://github.com/axmolengine/glslcc/releases/download/v1.9.3/glslcc-1.9.3-osx.tar.gz to /Users/chadwyk/xcode_projects/axmol-dev/tools/external/glslcc-osx.tar.gz ...
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
100 1821k  100 1821k    0     0  4030k      0 --:--:-- --:--:-- --:--:-- 4030k
build1k: Using glslcc: /Users/chadwyk/xcode_projects/axmol-dev/tools/external/glslcc/glslcc, version: 1.9.3
build1k: Downloading https://cmake.org/files/dev/cmake-3.27.20230808-g94eae40-macos-universal.tar.gz to /Users/chadwyk/xcode_projects/axmol-dev/tools/external/cmake-3.27.20230808-g94eae40-macos-universal.tar.gz ...
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100 74.3M  100 74.3M    0     0  1585k      0  0:00:48  0:00:48 --:--:-- 1861k
Password:
build1k: Deleting /Users/chadwyk/xcode_projects/axmol-dev/tools/external/cmake-3.27.20230808-g94eae40-macos-universal ...
build1k: Using cmake: /Users/chadwyk/xcode_projects/axmol-dev/tools/external/cmake/bin/cmake, version: 3.27.20230808-g94eae40
build1k: setup successfully, please restart the terminal to make added system variables take effect

**Reply by halx99** [reply]
i guess maybe issue caused by the prebuilt tool glslcc, it's only build x64, maybe m1 can't execute it correctly?

**Comment by halx99** [comment]
yes

获取Outlook for Android<https://aka.ms/AAb9ysg>
________________________________
